### PR TITLE
zig: update to 0.12.1.

### PR DESCRIPTION
**Pull request by OliveThePuffin**
#### Testing the changes
- I tested the changes in this PR: YES

#### Local build testing
- I built this PR locally for my native architecture, (ARCH-LIBC) x86_64-glibc

### CODE REVIEW

**Comment by OliveThePuffin** [comment]
This should close https://github.com/void-linux/void-packages/issues/45424

File: srcpkgs/zig/template [hunk 1]
```diff
@@ -1,20 +1,20 @@
 # Template file for 'zig'
 pkgname=zig
-version=0.10.1
-revision=3
+version=0.12.0
+revision=1
 archs="x86_64* aarch64*"
 build_style=cmake
-configure_args="-DZIG_TARGET_MCPU=baseline"
+configure_args="-DZIG_TARGET_MCPU=baseline -DZIG_STATIC_ZLIB=on"
```
Comment (OliveThePuffin) [inline srcpkgs/zig/template:7]: Had to do this because of https://github.com/ziglang/zig/issues/12069

**Comment by dkwo** [comment]
Did you try building anything with it? Hopefully soon river or waylock will tag a release using zig 0.12..

**Comment by OliveThePuffin** [comment]
Yeah. I built some of my advent of code stuff which works fine. I also built https://github.com/andrewrk/poop with it and it works without a hitch

**Comment by icp1994** [comment]
You need to rebuild (aka revbump) packages in the repo which which require zig for building and commit them within this PR. See https://github.com/void-linux/void-packages/pull/49436 as an example for how this workflow.

**Comment by OliveThePuffin** [comment]
Oh, I see what you mean now.

So 
river, waylock, ncdu2, zls, and linuxwave (templates that use zig-build)
all must build in this repo before merging. Makes sense.

I'll try to bump them now and see, but due to the breaking nature of early zig, this may have to wait a while.

**Comment by tranzystorekk** [comment]
Did some tests, our `zig-build` build_style needs to be ported to use `-Doptimize=ReleaseSafe` for setting profile, but ncdu2 2.4 also sets a preferred optimize level which replaces this flag with `-Drelease=<bool>` altogether :/

**Comment by OliveThePuffin** [comment]
I'm starting to think I bit off more than I could chew

**Comment by realcharmer** [comment]
@OliveThePuffin are you planning on finishing this up?

**Comment by OliveThePuffin** [comment]
I did, but right now the repo/xbps-src is breaking on me saying that "neither git or git are available in your system!" which is just false. I zapped and binary-bootstrapped, but nothing is fixing this. If you want to, feel free to take it away.

**Comment by Luciogi** [comment]
> I did, but right now the repo/xbps-src is breaking on me saying that "neither git or chroot-git are available in your system!" which is just false. I zapped and binary-bootstrapped, but nothing is fixing this. If you want to, feel free to take it away.

update master branch, and rebase this branch against master

**Comment by tranzystorekk** [comment]
river 0.3.0 won't support zig 0.12, we need to wait for 0.4.0, there seems to be some [work being done](https://codeberg.org/river/river/commits/branch/next-zig)

**Comment by dkwo** [comment]
New river and waylock are now out, which makes this PR viable.
However, the build style probably needs upgrading to use zig package manager, see e.g. https://codeberg.org/river/river/src/branch/master/PACKAGING.md

This works for me on x86_64 with static zig 0.12, but it probably needs some refinement to be included in void's build style:
```
 # Template file for 'river'
 pkgname=river
-version=0.2.4
-revision=3
+version=0.3.2
+revision=1
 archs="~i686* ~armv6l* ~armv7l*"
 build_style=zig-build
-configure_args="$(vopt_if xwayland -Dxwayland) -Dpie"
-hostmakedepends="pkg-config wayland-devel scdoc"
-makedepends="wlroots0.16-devel libevdev-devel pixman-devel
+hostmakedepends="pkg-config wayland-devel scdoc zig"
+makedepends="wlroots0.17-devel libevdev-devel pixman-devel
  wayland-protocols libxkbcommon-devel wayland-devel"
-depends="$(vopt_if xwayland xorg-server-xwayland)"
+depends="xorg-server-xwayland"
 short_desc="Dynamic tiling Wayland compositor"
 maintainer="Orphaned <orphan@voidlinux.org>"
 license="GPL-3.0-only"
 homepage="https://codeberg.org/river/river"
 distfiles="https://codeberg.org/river/river/releases/download/v${version}/river-${version}.tar.gz"
-checksum=26c1c41a65ce3804069afad6988410515cf478d2b76303ebc699766d3d4dc69f
+checksum=acb273487eb0cf9222c14e050c0faad692f426dba599b62e4c743ba43328ee9e
+_deps="https://codeberg.org/ifreund/zig-pixman/archive/v0.1.0.tar.gz
+ https://codeberg.org/ifreund/zig-wayland/archive/v0.1.0.tar.gz
+ https://codeberg.org/ifreund/zig-wlroots/archive/v0.17.0.tar.gz
+ https://codeberg.org/ifreund/zig-xkbcommon/archive/v0.1.0.tar.gz"
 
-build_options="xwayland"
-build_options_default="xwayland"
-desc_option_xwayland="Enable Xwayland support"
+post_fetch() {
+       for _d in $_deps; do
+               zig fetch --global-cache-dir $XBPS_BUILDDIR/zig-input $_d
+       done
+}
+
+do_build() {
+       DESTDIR="zig-out" zig build --system $XBPS_BUILDDIR/zig-input/p -Doptimize=ReleaseSafe -Dpie -Dxwayland --prefix /usr -Dcpu=baseline install
+}
+
+do_check() {
+       zig build --system $XBPS_BUILDDIR/zig-input/p test
+}
```

**Comment by dkwo** [comment]
Similarly:
```
 # Template file for 'waylock'
 pkgname=waylock
-version=0.6.2
+version=1.1.0
 revision=1
 build_style=zig-build
-configure_args="-Dpie"
 hostmakedepends="pkg-config wayland-devel scdoc"
 makedepends="libxkbcommon-devel pam-devel wayland-protocols wayland-devel"
 short_desc="Simple screenlocker for wayland compositors"
@@ -11,7 +10,19 @@ maintainer="Orphaned <orphan@voidlinux.org>"
 license="ISC"
 homepage="https://codeberg.org/ifreund/waylock"
 distfiles="https://codeberg.org/ifreund/waylock/releases/download/v${version}/waylock-${version}.tar.gz"
-checksum=4a794dedd1b11f3b9bb7b46f8a93c20aac5ac1d40a851cabaa7ae84fd36686fa
+checksum=fd200ddaf416e45a8dc6e212fb8b0fba2cf7bdcfaf5f65a361d8b0313600d02e
+_deps="https://codeberg.org/ifreund/zig-wayland/archive/v0.1.0.tar.gz
+ https://codeberg.org/ifreund/zig-xkbcommon/archive/v0.1.0.tar.gz"
+
+post_fetch() {
+       for _d in $_deps; do
+               zig fetch --global-cache-dir $XBPS_BUILDDIR/zig-input $_d
+       done
+}
+
+do_build() {
+       DESTDIR="zig-out" zig build --system $XBPS_BUILDDIR/zig-input/p -Doptimize=ReleaseSafe -Dpie --prefix /usr -Dcpu=baseline install
+} 
```

**Comment by tranzystorekk** [comment]
since the zig pm scheme behaves similar to cargo we could just configure global cache in the build style and et voila

**Comment by dkwo** [comment]
makes sense

**Comment by tranzystorekk** [comment]
Just tested for waylock and river, adding `--global-cache-dir /host/zig` to zig-build style works well

**Comment by realcharmer** [comment]
Thanks for the heads-up, I'll update both PRs tomorrow.

**Comment by jason1987d** [comment]
Seeing river 0.3.2 uses zig 0.12 now. Also OP, do not get discouraged. Void packaging is not for the faint of heart. You will get fail, many times. The maintainers here have put me through eternal bootcamp with attitude to match.

**Comment by OliveThePuffin** [comment]
rebase went wrong 💀

**Comment by tranzystorekk** [comment]
I prepared build_style changes that should make package templates not require custom do_install hacks: https://github.com/tranzystorekk/void-packages/commit/69fed2ebf4b6eb2fb86053b588234d7911a30287

**Comment by OutOfContainment** [comment]
Quote from Andrew:
> I'm planning to release 0.13.0 and 0.12.1 on Monday, primarily featuring the LLVM 18 upgrade and musl 1.2.5 upgrade. If there are any other impactful changes you want to land, you have 5 days to get them in. 0.12.1 will be the last release in the 0.12.x series.
> the release is primarily aimed at helping distributions package zig since it updates to LLVM 18. distros which have llvm 17 can package zig 0.12.1; distros that have llvm 18 can package 0.13.0

**Comment by dkwo** [comment]
@tranzystorekk Great! I can confirm that works for me, with the trivial update for waylock, and this one for river:
```
# Template file for 'river'
 pkgname=river
-version=0.2.4
-revision=3
+version=0.3.2
+revision=1
 archs="~i686* ~armv6l* ~armv7l*"
 build_style=zig-build
-configure_args="$(vopt_if xwayland -Dxwayland) -Dpie"
+configure_args="-Dpie -Dxwayland"
 hostmakedepends="pkg-config wayland-devel scdoc"
-makedepends="wlroots0.16-devel libevdev-devel pixman-devel
+makedepends="wlroots0.17-devel libevdev-devel pixman-devel
  wayland-protocols libxkbcommon-devel wayland-devel"
-depends="$(vopt_if xwayland xorg-server-xwayland)"
+depends="xorg-server-xwayland"
 short_desc="Dynamic tiling Wayland compositor"
 maintainer="Orphaned <orphan@voidlinux.org>"
 license="GPL-3.0-only"
 homepage="https://codeberg.org/river/river"
 distfiles="https://codeberg.org/river/river/releases/download/v${version}/river-${version}.tar.gz"
-checksum=26c1c41a65ce3804069afad6988410515cf478d2b76303ebc699766d3d4dc69f
+checksum=acb273487eb0cf9222c14e050c0faad692f426dba599b62e4c743ba43328ee9e
 
-build_options="xwayland"
-build_options_default="xwayland"
-desc_option_xwayland="Enable Xwayland support"
+do_check() {
+       zig build --system /host/zig/p test
+}
```
(The reason for removing the xwayland option is that, even if build with it, it can be disabled at runtime.)

**Comment by icp1994** [comment]
Shouldn't `--global-cache-dir` be something like `${XBPS_HOSTDIR}/zigcache` rather than `/host/zig`?

**Comment by tranzystorekk** [comment]
i took it from the cargo/rust style, afaiu its symlinked anyway

**Comment by OliveThePuffin** [comment]
I'm assuming based on the logs of the build that the failed river compile is a source code issue

**Comment by Luciogi** [comment]
zig 0.12.1 is available

**Comment by tranzystorekk** [comment]
> I'm assuming based on the logs of the build that the failed river compile is a source code issue

no, that's my bad, search prefix should be `--search-prefix "${XBPS_CROSS_BASE}/usr"`

**Comment by OliveThePuffin** [comment]
Zig, river, and waylock should all be updated now. I even minimally tested some functionality on my machine and it works.

Now, there are a few other packages that depend on zig that seem to not have a trivial update:

linuxwave - still on just 0.10.1 - maybe we ditch it?
zls - comes with software from 3 repos. I can update the main one, but I'm not so sure about the others
ncdu2 - seems like it would be trivial but fails at do-patch for some reason.

Thoughts?

**Comment by tranzystorekk** [comment]
My tested changes:

- ncdu2: https://github.com/tranzystorekk/void-packages/commit/5843ee4b885aeb8a9c7e56343e9c2aa0c0f046f2
- zls: https://github.com/tranzystorekk/void-packages/commit/9086dcbac5209c40d15ea6954927f9a82ddefe8e

linuxwave PR you mentioned should work already, the only wip is making its CI work with newer zig

**Comment by dataCobra** [comment]
There is zig 0.13.0 out since 07.06.2024.

`river` and `waylock` are both available in new versions with 0.13.0 support.

**Comment by OliveThePuffin** [comment]
> My tested changes:
> 
>     * ncdu2: [tranzystorekk@5843ee4](https://github.com/tranzystorekk/void-packages/commit/5843ee4b885aeb8a9c7e56343e9c2aa0c0f046f2)
> 
>     * zls: [tranzystorekk@9086dcb](https://github.com/tranzystorekk/void-packages/commit/9086dcbac5209c40d15ea6954927f9a82ddefe8e)
> 
>     * linuxwave: the PR you mentioned should work already, the only wip is making its CI work with newer zig

bro, you are a rock star. I just added your fork as a remote and I'm using those changes for this PR now.

**Comment by OutOfContainment** [comment]
IMHO it seems like Orhun is too busy with other stuff to pay any attention to linuxwave, so we could:
 - ditch it as @OliveThePuffin suggested
 - push a fork based on [this](https://github.com/orhun/linuxwave/pull/28) pr
 
To be honest, I don't see any usecase for linuxwave other than a throwaway toy, so it kinda seems like a burden to Void right now.

**Comment by tranzystorekk** [comment]
I'm fine with either dropping linuxwave or attempting to make it work, but maintaining a fork seems excessive for a toy package.

I'd also consider leaving it as a known FTBFS for now since it doesn't affect other packages or most users likely.

**Comment by OliveThePuffin** [comment]
What is FTBFS?

**Comment by tranzystorekk** [comment]
> What is FTBFS?

Fails To Build From Source

**Comment by tranzystorekk** [comment]
Superseded by #51223

Thank you for the discussion in this PR; I decided to take the changes for 0.13 that are already a basis for this PR and open a new PR instead.

**Pull request closed without merging**